[WIP] Introduce Hosted Support

tduffield opened the pull request:
This PR is based on #181. We'll want to merge that PR and then rebase this one. 

--- 

Hosted support optimizes the behavior of Dobi for a hosted CI
environment such as Jenkins, TravisCI, CircleCI, or Buildkite.

Hosted Mode changes two key behaviors:

1. We use `git` rather than the file modification time to determine when
   a file has last been modified.
2. Since the Dobi records in `.dobi` need to be stored outside of the
   workspace, the `LastModified` value is kept inside the record, rather
   than being determined based on the file modification time.

To enable Hosted Mode, simply set `DOBI_HOSTED` to any value. You can
also set `meta.hosted` to `true` in your Dobi YAML.

tduffield commented:
Alright, fully rebased with unit and functional tests!

dnephin commented:
I'm not sure when I will be able to give it a full review and try it out, but I wanted to say that it looks pretty cool.

Thanks for working on it!

tduffield commented:
@dnephin I've rebased and updated my PR based on your most recent changes. I'd love to see if we can get this into the upstream so I can avoid having to maintain a fork. Thanks!
